# Practice cleaning missing values with California Cities Report

In this project, we will practice how to identifying and cleaning missing (null) values. For this project, we define "missing" only as np.nan values. First, we will practice how to find the missing values, drop columns or rows having missing values, and then we will practice different methods of filling the missing values. We will cover other missing values (for example, empty strings, etc) in other projects.

**Our main focus will be on the pandas methods for:**

1. Identification and Cleaning Missing Values:

    - ``.info()``
    - ``.isna()``
    - ``.isnull()``
    - ``.dropna()``

2. Data Imputation Methods:

    - ``.fillna()``
    - ``.bfill()``
    - ``.ffill()``
    - ``.mode()``
    - ``.mean()``
    - ``.median()``
    
Lets jump right into our notebook and run all the cells till the first question. For this activity, we will be using an interesting dataset of California Cities which unfortunately has been corrupted due to human error and needs some fixing for the missing values. This dataset contains very important information about 482 cities in the Califorina state with data about its cities, their latitude, logitude, elevation, populations, land area, and water area in different metric units. Lets help clean this data so the scientists can use it for geological research.

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import pandas as pd

In [45]:
df = pd.read_csv('files/california_cities.csv')
df.head()

city       latd       longd  elevation_m  elevation_ft  \
0     Adelanto  34.576111 -117.432778        875.0        2871.0   
1  AgouraHills  34.153333 -118.761667        281.0         922.0   
2      Alameda  37.756111 -122.274444          NaN          33.0   
3       Albany  37.886944 -122.297778          NaN          43.0   
4     Alhambra  34.081944 -118.135000        150.0         492.0   

   population_total  area_total_sq_mi  area_land_sq_mi  area_water_sq_mi  \
0           31765.0            56.027           56.009             0.018   
1           20330.0             7.822            7.793             0.029   
2           75467.0            22.960           10.611            12.349   
3           18969.0             5.465            1.788             3.677   
4           83089.0             7.632            7.631             0.001   

   area_total_km2  area_land_km2  area_water_km2  area_water_percent  
0         145.107        145.062           0.046                0.03  
1          20.260         20.184           0.076                0.37  
2          59.465         27.482          31.983                 NaN  
3          14.155          4.632           9.524                 NaN  
4          19.766         19.763           0.003                 NaN

In [46]:
df.columns

Index(['city', 'latd', 'longd', 'elevation_m', 'elevation_ft',
       'population_total', 'area_total_sq_mi', 'area_land_sq_mi',
       'area_water_sq_mi', 'area_total_km2', 'area_land_km2', 'area_water_km2',
       'area_water_percent'],
      dtype='object')

1. Find out the number of missing values in each column and store them in the variable ``col_missing_values``

    Make sure you run all the previous cells. Don't worry if you screw up with the DataFrame! just reload it with the first line of the notebook.

In [47]:
df.isnull().sum()

city                   0
latd                  12
longd                 15
elevation_m           50
elevation_ft          14
population_total       2
area_total_sq_mi       6
area_land_sq_mi        3
area_water_sq_mi       5
area_total_km2         7
area_land_km2          6
area_water_km2         6
area_water_percent    62
dtype: int64

5. Find the total missing values in the whole dataset and store the number in ``df_missing_values``

    You must modify the df variable itself. Don't worry if you screw up with the DataFrame! just reload it with the first line of the notebook.

In [48]:
df_missing_values = df.isnull().sum().sum()
df_missing_values

188

6. Drop the column ``area_water_percent`` as it has the most of its values missing

    You have to drop this column permanently as we can not use it for any purpose.

In [49]:
df.drop(columns=['area_water_percent'], inplace=True)

7. Drop the rows having missing values and store the resulting DataFrame in the variable ``df_narows_dropped``

In [50]:
df_narows_dropped = df.dropna()

In [51]:
df_narows_dropped = df[df.notna().all(axis=1)]

In [52]:
df_narows_dropped = df.dropna(axis=0)

8. Drop Rows with More Than 5 Missing Values

    Remove rows from the DataFrame where the count of non-null values is less than 5, effectively eliminating rows with more than 5 missing values. Save the resulting DataFrame in the variable ``df_rows_dropped`` and make use of the ``thresh`` parameter for this operation.

In [53]:
df_rows_dropped = df.dropna(thresh=5)

In [54]:
# Generate a boolean mask for rows having 5 or fewer missing values
row_mask = df.count(axis=1) >= df.shape[1] - 5

# Produce a new dataframe composed solely of rows passing the mask
df_rows_dropped = df[row_mask]

9. Remove Columns with Missing Values

    Eliminate the columns in the DataFrame that contain missing values. Store the resulting DataFrame in the variable ``df_nacols_dropped``.

In [55]:
df_nacols_dropped = df.dropna(axis = 1)

10. Drop columns with more than 10 missing values and store the resulting DataFrame in the variable ``df_cols_dropped``

In [56]:
missing_values = df.isnull().sum()
missing_values

city                 0
latd                12
longd               15
elevation_m         50
elevation_ft        14
population_total     2
area_total_sq_mi     6
area_land_sq_mi      3
area_water_sq_mi     5
area_total_km2       7
area_land_km2        6
area_water_km2       6
dtype: int64

In [57]:
df.loc[: , missing_values <= 10]

city  population_total  area_total_sq_mi  area_land_sq_mi  \
0       Adelanto           31765.0            56.027           56.009   
1    AgouraHills           20330.0             7.822            7.793   
2        Alameda           75467.0            22.960           10.611   
3         Albany           18969.0             5.465            1.788   
4       Alhambra           83089.0             7.632            7.631   
..           ...               ...               ...              ...   
477   Yountville            2933.0             1.531            1.531   
478        Yreka            7765.0            10.053            9.980   
479     YubaCity           64925.0            14.656           14.578   
480      Yucaipa           51367.0            27.893           27.888   
481  YuccaValley           20700.0            40.015           40.015   

     area_water_sq_mi  area_total_km2  area_land_km2  area_water_km2  
0               0.018         145.107        145.062           0.046  
1               0.029          20.260         20.184           0.076  
2              12.349          59.465         27.482          31.983  
3               3.677          14.155          4.632           9.524  
4               0.001          19.766         19.763           0.003  
..                ...             ...            ...             ...  
477             0.000           3.966          3.966           0.000  
478             0.073          26.036         25.847           0.188  
479             0.078          37.959         37.758           0.201  
480             0.005          72.244         72.231           0.013  
481             0.000         103.639        103.639           0.000  

[482 rows x 8 columns]

In [58]:
df_cols_dropped = df.loc[: , missing_values <= 10]

## Data Imputation Methods

11. Fill the 50 missing values in ``elevation_m`` with ``-999``. Store your result in the variable ``filled_elevation_m``



In [59]:
filled_elevation_m = df['elevation_m'].fillna(-999)

In [60]:
filled_elevation_m

0       875.0
1       281.0
2      -999.0
3      -999.0
4       150.0
        ...  
477      30.0
478     787.0
479      18.0
480     798.0
481    1027.0
Name: elevation_m, Length: 482, dtype: float64

12. Fill the 7 missing values in ``area_total_km2`` with the value 0 permanently, store your result in the variable ``filled_area_total``

In [63]:
filled_area_total = df['area_total_km2'].fillna(0, inplace = True)

Second solution

In [ ]:
df.loc[df['area_total_km2'].isna(), 'area_total_km2'] = 0
filled_area_total = df['area_total_km2']

13. Fill the missing values of the column ``latd`` using backward filling method and store your result in the variable ``bfill_latd``

In [64]:
bfill_latd = df['latd'].bfill()

Second solution

In [ ]:
bfill_latd = df['latd'].fillna(method='bfill')

14. Fill the 15 missing values of the column ``longd`` using forwarding filling method and store your result in the variable ``ffill_longd``

In [66]:
ffill_longd = df['longd'].ffill()

15. Fill the 2 missing values of the column ``population_total`` with the mean of the column and store your result in the variable ``mean_total_population``

In [67]:
mean_total_population = df['population_total'].fillna(value = df['population_total'].mean())

16. Fill the 5 missing values of the column ``area_water_sq_mi`` with the median value of the column and store your result in the variable ``median_fill``

In [68]:
median_fill = df['area_water_sq_mi'].fillna(value = df['area_water_sq_mi'].median())

17. Fill the 6 missing values of the column ``area_land_km2`` with the ``mode`` value of the column and store your result in the variable ``mode_fill``

In [71]:
mode_fill = df['area_land_km2'].fillna(df['area_land_km2'].mode()[0])

# Cleaning duplicate data from an Online Retail store

In this project, we will be working with an Online Retail dataset, which contains information about transactions made by an online retailer. The dataset includes details such as the invoice number, stock code, description, quantity, invoice date, unit price, customer ID, and country.

Duplicate data is a common issue that can arise in datasets like these, and it's important to be able to identify and remove these duplicates in order to avoid biased or inaccurate analyses. In this practice lab, we will be focusing on identifying and dropping duplicate data using the ``duplicated()`` and ``drop_duplicates()`` methods in pandas.

We will start by exploring the dataset and identifying any duplicate rows using the ``duplicated()`` method. Once we have identified the duplicates, we will use the ``drop_duplicates()`` method to remove them from the dataset. Also, the practice includes usage of different parameters that can be used with the ``drop_duplicates()`` method, such as keep and subset.

By the end of this practice lab, you should feel comfortable using pandas to identify and remove duplicate data from a dataset, and have a better understanding of how to use the ``duplicated()`` and ``drop_duplicates()`` methods with different parameters to achieve the desired outcome.

In [1]:
#to read the dataframe
import pandas as pd

In [2]:
# Reading dataframe
df = pd.read_csv('files/OnlineRetail.csv')

In [3]:
df.head()

InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

      InvoiceDate  UnitPrice  CustomerID         Country  
0  12/1/2010 8:26       2.55     17850.0  United Kingdom  
1  12/1/2010 8:26       3.39     17850.0  United Kingdom  
2  12/1/2010 8:26       2.75     17850.0  United Kingdom  
3  12/1/2010 8:26       3.39     17850.0  United Kingdom  
4  12/1/2010 8:26       3.39     17850.0  United Kingdom

In [4]:
#show the number of cells in the dataframe
print("dataset size: ", df.size)

#show the number of records (rows) in the dataframe
print("number of talks: ", len(df))

#show the number of features (coulmns) in the dataframe
print("number of features: ", len(df.columns)) 

dataset size:  4335272
number of talks:  541909
number of features:  8


In [5]:
df.dtypes

InvoiceNo       object
StockCode       object
Description     object
Quantity         int64
InvoiceDate     object
UnitPrice      float64
CustomerID     float64
Country         object
dtype: object

1. **Which of the following parameters is used to only consider certain columns for identifying duplicates and it by default uses all of the columns?**

``subset``

2. **Which of the following parameters is used to determine whether to modify the DataFrame rather than creating a new one?**

``inplace``

3. **Which of the following parameters takes 'first' as a value?**

`keep`

The solution is ``keep``. It means "determine which duplicates (if any) to keep. For example, ``first`` will drop duplicates except for the first occurrence".

## Basic Activities

For most activities, make sure you are not modifying the original dataframe. In other words: either don't pass an ``inplace`` parameter, or pass it as ``inplace=False``, so your modifications return a new variable.

4. Select duplicate rows in a dataframe from the dataset?

    Perform the selection and store the results in the variable ``duplicate_rows``.

    - Note: use the default parameter of ``keep='first'``.

In [8]:
duplicate_rows = df[df.duplicated(keep='first')]

5. What is the number of duplicate rows?

In [10]:
len(duplicate_rows)

5268

6. Find and drop duplicate rows based on InvoiceNo, StockCode, Quantity, and UnitPrice columns

    This data contains duplicate orders with the same quantity and unit price, so drop these duplicates.

    Perform the dropping and store the results in the variable ``df_without_duplicate_orders``.

In [18]:
df_without_duplicate_orders = df.drop(df.loc[df[['InvoiceNo', 'StockCode','Quantity','UnitPrice']].duplicated()].index)

In [ ]:
df_without_duplicate_orders = df.drop_duplicates(subset=['InvoiceNo', 'StockCode', 'Quantity', 'UnitPrice'], inplace=False)

7. Drop duplicates while keeping the first non-NaN value based on InvoiceNo, StockCode, and CustomerID columns

    As each invoice should have the stock code only one time for each customer and the customer may have different quantities: Drop duplicates while keeping the first non-NaN value.

    Perform the dropping and store the results in the variable ``df_keep_first``.

In [24]:
df_keep_first = df.drop_duplicates(subset=['InvoiceNo', 'StockCode', 'CustomerID'], keep='first')
df_keep_first

InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
541904    581587     22613          PACK OF 20 SPACEBOY NAPKINS        12   
541905    581587     22899         CHILDREN'S APRON DOLLY GIRL          6   
541906    581587     23254        CHILDRENS CUTLERY DOLLY GIRL          4   
541907    581587     23255      CHILDRENS CUTLERY CIRCUS PARADE         4   
541908    581587     22138        BAKING SET 9 PIECE RETROSPOT          3   

            InvoiceDate  UnitPrice  CustomerID         Country  
0        12/1/2010 8:26       2.55     17850.0  United Kingdom  
1        12/1/2010 8:26       3.39     17850.0  United Kingdom  
2        12/1/2010 8:26       2.75     17850.0  United Kingdom  
3        12/1/2010 8:26       3.39     17850.0  United Kingdom  
4        12/1/2010 8:26       3.39     17850.0  United Kingdom  
...                 ...        ...         ...             ...  
541904  12/9/2011 12:50       0.85     12680.0          France  
541905  12/9/2011 12:50       2.10     12680.0          France  
541906  12/9/2011 12:50       4.15     12680.0          France  
541907  12/9/2011 12:50       4.15     12680.0          France  
541908  12/9/2011 12:50       4.95     12680.0          France  

[531225 rows x 8 columns]

8. Drop duplicates while keeping the last order based on StockCode and InvoiceWeekday columns

    If you want to show number of unique transactions per weekday and StockCode combination, you will need to drop duplicate stockcode on same day.

    Perform the dropping and store the results in the variable ``df_unique_stock_day``.

    - Note: use the ``df_weekday`` prepared for you to solve this activity.

In [26]:
# Convert InvoiceDate column to datetime
df_weekday = df

df_weekday['InvoiceDate'] = pd.to_datetime(df['InvoiceDate'])

# Create a new column for weekday
df_weekday['InvoiceWeekday'] = df['InvoiceDate'].dt.weekday

In [29]:
# Drop duplicates based on StockCode and InvoiceWeekday

df_unique_stock_day = df_weekday.drop_duplicates(subset=['StockCode', 'InvoiceWeekday'], keep='last', inplace=False)

9. Drop all duplicate invoices

    Imagine it is black friday and each customer is allowed to buy only at one invoice. So, we need to drop all the duplicate invoices.

    Perform the dropping and store the results in the variable ``df_black_friday``.

In [36]:
df_black_friday = df.drop_duplicates(subset=['InvoiceNo'], keep=False, inplace=False)

10. Drop duplicate countries while keeping the first row

    Imagine we want to know all unique countries in our stock, drop duplicate countries keeping first row.

    Perform the dropping and store the results in the variable ``df_unique_countries``.

In [38]:
df_unique_countries = df.drop_duplicates(subset=['Country'], keep='first', inplace=False)

11. Drop duplicate products while keeping last based on StockCode, Description, and UnitPrice

    Imagine we want to know all ordered products in our retail, drop duplicate products based on StockCode, Description, and UnitPrice.

    Perform the dropping and store the results in the variable ``df_unique_products``.

In [40]:
df_unique_products = df.drop_duplicates(subset=['StockCode','Description','UnitPrice'], keep='last', inplace=False)

12. Drop all duplicate rows based on TotalCost and CustomerID while keeping first

    We want to know all unique total costs paid by each different customer, So drop these duplicates.

    Perform the dropping and store the results in the variable ``df_customer_unique_payments``.

In [45]:
df_cost = df
# Create a new column for TotalCost
df_cost['TotalCost'] = df_cost['Quantity']*df_cost['UnitPrice']

In [46]:
df_customer_unique_payments = df_cost.drop_duplicates(subset=['UnitPrice','TotalCost'], keep='first', inplace=False)

13. Drop all duplicate rows while keeping first

    Perform the dropping and store the results in the variable ``df_unique``.

In [47]:
df_unique = df.drop_duplicates(keep='first')

# Practice Data Cleaning and String Handling with City Bike data

In New York, Citi Bike operates a network of bicycle rental stations with a subscription system. The stations are all over the city and provide a convenient way to get around. Bike rental data is made public in an anonymized form and has been analyzed in various ways. The data has many string columns to be modified before further usage.

In this section we will practice how to interact and manipulate string data using Pandas ``.str`` method. We will cover the most common string methods like:

- ``.str.capitalize(), .str.lower(), .str.upper(), .str.title()``
- ``.str.contains()``
- ``.str.count()``
- ``.str.endswith(), .str.startswith()``
- ``.str.find()``
- ``.str.join()``
- ``.str.len()``
- ``.str.replace()``
- ``.str.split()``
- ``.str.isdigit() .str.isalpha()``

Let's Jump into the practice so that we will clean the data and bring the columns with messy string forms to nice and usable format.

### Quick analysis

The first few cells in the Notebook read the dataset, display the first few rows (``df.head()``) and show you the column types (``df.info()``). Take your time, exploring the dataset and the columns. Remember that the focus of this project is String Manipulation, so focus on those columns that are of type object.

Once you're ready, jump to the next page to start the activities!

In [1]:
import pandas as pd

In [7]:
df = pd.read_csv('files/citi-bikes.csv')
df

first name  last name  tripduration  start station id start station name  \
0         misty    HarRISs          3117               301  E 2 St & Avenue B   
1        Alexis   Thompson           690               301  E 2 St & Avenue B   
2          Jodi    Sanders           727               301  E 2 St & Avenue B   
3       matrick      Evans           698               301  E 2 St & Avenue B   
4        Amanda   DaniEEls           351               301  E 2 St & Avenue B   
...         ...        ...           ...               ...                ...   
4207     Joshua  Donaldson          1225               301  E 2 St & Avenue B   
4208     Deanna     Nelson           939               301  E 2 St & Avenue B   
4209   Jennifer    Sanchez           914               301  E 2 St & Avenue B   
4210     Steven      Riley          1581               301  E 2 St & Avenue B   
4211     Taylor    Hancock           247               301  E 2 St & Avenue B   

      end station id          end station name  bikeid    usertype  \
0                301         E 2 St & Avenue B   18070  Subscriber   
1                349   Rivington St & Ridge St   19699  Subscriber   
2               2010      Grand St & Greene St   20953  Subscriber   
3                527           E 33 St & 2 Ave   23566  Subscriber   
4                250  Lafayette St & Jersey St   17545  Subscriber   
...              ...                       ...     ...         ...   
4207             442           W 27 St & 7 Ave   15859    Customer   
4208             251       Mott St & Prince St   15683    Customer   
4209             251       Mott St & Prince St   20219    Customer   
4210             465        Broadway & W 41 St   23176  Subscriber   
4211             511        E 14 St & Avenue B   23484  Subscriber   

      birth year  gender                       emails         pin  
0         1986.0       1    osborneangela@example.com  JPCeiknPMw  
1         1985.0       1          sbutler@example.net  PSxNqDPDWe  
2         1982.0       1          egarcia@example.net  pxTMwm166p  
3         1976.0       1           zmoore@example.com  w5dktoeXrq  
4         1959.0       1        derrick11@example.com  8hHtmBR0Tt  
...          ...     ...                          ...         ...  
4207         NaN       0          kelly26@example.net  NaaPhq7R38  
4208         NaN       0           seth26@example.net  PWs7OnO37x  
4209         NaN       0         tamara26@example.org  NNEAV3nI8i  
4210      1996.0       2  carlosstevenson@example.com  dGBYIsZzM2  
4211      1982.0       1     arnolddaniel@example.com  v2VFozNw7y  

[4212 rows x 13 columns]

In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4212 entries, 0 to 4211
Data columns (total 13 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   first name          4212 non-null   object 
 1   last name           4212 non-null   object 
 2   tripduration        4212 non-null   int64  
 3   start station id    4212 non-null   int64  
 4   start station name  4212 non-null   object 
 5   end station id      4212 non-null   int64  
 6   end station name    4212 non-null   object 
 7   bikeid              4212 non-null   int64  
 8   usertype            4212 non-null   object 
 9   birth year          3682 non-null   float64
 10  gender              4212 non-null   int64  
 11  emails              4212 non-null   object 
 12  pin                 4212 non-null   object 
dtypes: float64(1), int64(5), object(7)
memory usage: 427.9+ KB


1. **Capitalize the column first name**

    If you explore the DataFrame, you'll see that the column ``first name`` is "inconsistent" with its capitalization. Some names are capitalized (Alexis, Jodi), but some others are not (misty, matrick).

    Create a new series ``capital_first_name`` that contains the results of the column first name correctly capitalized.

In [8]:
capital_first_name = df['first name'].str.capitalize()

In [9]:
capital_first_name

0          Misty
1         Alexis
2           Jodi
3        Matrick
4         Amanda
          ...   
4207      Joshua
4208      Deanna
4209    Jennifer
4210      Steven
4211      Taylor
Name: first name, Length: 4212, dtype: object

2. Make the Column last name as lower case

    Now we can see the Column ``last name`` has very messy format too. Some of the middle letters are capitalized like in ``HarRISs``, ``DaniEEIs`` and some of the first names are not capitalize. So, convert all of them to lower case and store the result in the variable ``lower_last_name``

In [10]:
lower_last_name = df['last name'].str.lower()

In [11]:
lower_last_name

0         harriss
1        thompson
2         sanders
3           evans
4        danieels
          ...    
4207    donaldson
4208       nelson
4209      sanchez
4210        riley
4211      hancock
Name: last name, Length: 4212, dtype: object

3. Make ``last name`` as Upper case. Store your answer in the variable ``upper_last_name``

In [12]:
upper_last_name = df['last name'].str.upper()

4. How many users in the Column ``usertype`` are Customer

    Lets count all the Customers in the column ``usertype`` and sum them up. Store your sum in the ``customer_counts`` variable.

In [15]:
customer_df = df[df['usertype'] == 'Customer']
customer_counts = len(customer_df['emails'].unique())

530

Solution 1

In [16]:
customer_counts = df['usertype'].str.count('Customer')
customer_counts=customer_counts.sum()
customer_counts

530

Solution 2

In [ ]:
customer_counts = df['usertype'].value_counts()
customer_counts

5. How many users in the Column ``usertype`` are ``Subscribers``

    As you got the total number of Customers in ``usertype`` from the previous question, then also find how many Subscribers are there in total.

    You can subtract the number of Customers from the total lenght of the dataframe to find the remain - which are Subscribers.

In [19]:
df['usertype'].str.count('Subscriber').sum()

3682

6. Find the words in Column ``pin`` which contain the substring ``lol`` and store your selection in the variable ``word_having_lols``

In [24]:
word_having_lols = df['pin'].str.contains('lol')
df['pin'][word_having_lols]

59        Vtdqzzlolk
65          foundlol
117       loVulolKDF
167       nlolzlVd23
2656      6Kolol9toS
2660      lolLQigMjv
2679        FounDlol
2729      26lolbYZ66
2744      PlolJhPbGQ
2765     UuwlolrTHDW
2783      kzlolZJvRk
2804      kfJAlolWxZ
2828      Eh8lolEwKd
2951    145loleGTrpG
Name: pin, dtype: object

7. Find the names in the Column ``first name`` which start with the letter Z

    Find all the names in the ``first name`` column that start with the latter "Z". Store the result in the variable ``starts_with_z``.

    >***Be careful!*** *It's capital Z, not lowercase z.*

In [26]:
starts_with_z = df['first name'].str.startswith('Z')
df['first name'][starts_with_z]

233     Zachary
348     Zachary
431     Zachary
651     Zachary
847     Zachary
1035    Zachary
1270    Zachary
1366    Zachary
1451    Zachary
2371    Zachary
2508    Zachary
2589    Zachary
2854    Zachary
2964    Zachary
3128    Zachary
3156    Zachary
3535    Zachary
3581    Zachary
3712    Zachary
3814    Zachary
Name: first name, dtype: object

8. How many first names start with the letter 'Z'

In [29]:
len(df['first name'][starts_with_z])

20

9. Find the names in the Column ``last name`` which end with 't' and store your result in the variable ``ends_with_t``

In [31]:
ends_with_t = df['last name'].str.endswith('t')
df['last name'][ends_with_t]

67      Lambert
101     Elliott
111      Knight
117     Burnett
126       Grant
         ...   
4107     Hebert
4140     Bryant
4159    Stewart
4181    Garrett
4183     Wright
Name: last name, Length: 145, dtype: object

10. How many Values in the Column last name end with 't'

In [32]:
ends_with_t.sum()

145

11. Concatenate 'bikeid' Entries with Space Separator

    Utilize the ``str.join()`` method to concatenate the bikeid values in the ``bikeid`` column, using a space as the separator. Store the result in the variable ``spaced_bikeids``.

In [37]:
df['bikeid']=df['bikeid'].astype(str)

spaced_bikeids = df['bikeid'].str.join(sep = ' ')

In [38]:
spaced_bikeids

0       1 8 0 7 0
1       1 9 6 9 9
2       2 0 9 5 3
3       2 3 5 6 6
4       1 7 5 4 5
          ...    
4207    1 5 8 5 9
4208    1 5 6 8 3
4209    2 0 2 1 9
4210    2 3 1 7 6
4211    2 3 4 8 4
Name: bikeid, Length: 4212, dtype: object

12. Create a new Column named ``name length`` having all the lengths of names from the Column ``first name``

In [43]:
df['name length'] = df['first name'].str.len()

13. Find if the Column ``pin`` is alpha numeric or it contains digits only

In [47]:
print(f"La columna pin es numerica? {df['pin'].str.isdigit()[0]}")
print(f"La columna pin es alfanumerica? {df['pin'].str.isalpha()[0]}")

La columna pin es numerica? False
La columna pin es alfanumerica? True


14. Verify if the Column ``tripduration`` has any non-numeric values or it contains digits only

In [51]:
df['tripduration'] = df['tripduration'].astype(str)

print(f"La columna tripduration es numerica? {df['tripduration'].str.isdigit()[0]}")
print(f"La columna tripduration es alfanumerica? {df['tripduration'].str.isalpha()[0]}")

La columna tripduration es numerica? True
La columna tripduration es alfanumerica? False


15. Check if any name in the Column ``first name`` has digit(s) or number(s) in it

In [55]:
print(f"La columna first name es numerica? {df['first name'].str.isdigit()[0]}")
print(f"La columna first name es alfanumerica? {df['first name'].str.isalpha()[0]}")
print(f"La columna first name es decimal? {df['first name'].str.isdecimal()[0]}")
df['first name'].str.isalpha().sum()

La columna first name es numerica? False
La columna first name es alfanumerica? True
La columna first name es decimal? False


4208

16. Split the emails in the emails column at ``@`` to find the Domain names and store them in the variable ``email_domains``

In [72]:
email_domains = df['emails'].str.split('@').str[1]

17. Replace the emails having ``.edu`` with ``.org`` and store the output in the variable ``edu_to_org``

In [73]:
# Replace .edu with .org in the email column
edu_to_org = df['emails'].str.replace('.edu', '.org',regex=True)

In [74]:
edu_to_org

0         osborneangela@example.com
1               sbutler@example.net
2               egarcia@example.net
3                zmoore@example.com
4             derrick11@example.com
                   ...             
4207            kelly26@example.net
4208             seth26@example.net
4209           tamara26@example.org
4210    carlosstevenson@example.com
4211       arnolddaniel@example.com
Name: emails, Length: 4212, dtype: object

18. Replace the numeric and the St values in ``end station name`` Column with ``<space>`` so that we can filter the address without street numbers. Store your result in the variable ``clean_address``

In [75]:
# Replace all numeric characters and 'St' with a space
clean_address = df['end station name'].str.replace(r'\d+|St', ' ', regex=True)

In [76]:
clean_address

0             E     & Avenue B
1        Rivington   & Ridge  
2           Grand   & Greene  
3                E     &   Ave
4       Lafayette   & Jersey  
                 ...          
4207             W     &   Ave
4208         Mott   & Prince  
4209         Mott   & Prince  
4210          Broadway & W    
4211          E     & Avenue B
Name: end station name, Length: 4212, dtype: object

# Matching Strings by Similarity using Levenshtein distance

In this project, we're going to resolve a very tough problem, which is identifying similar strings that are not completely equal. It's a very common and challenging data-cleaning task, so let's get right into it!

We have two CSVs containing company names. The problem is that the company names in these two CSVs don't match. Both have a single column, but if you explore them, you'll immediately spot the anomalies.

In [1]:
import itertools
import pandas as pd

# The new library!
from thefuzz import fuzz, process

In [2]:
df1 = pd.read_csv('files/companies_1.csv')
df2 = pd.read_csv('files/companies_2.csv')

In [3]:
df1.head(10)

CLIENT
0                    Adobe Systems, Inc.
1                       Adventist Health
2                                  AECOM
3   Aerojet Rockedyne Holdings (GenCorp)
4  Alameda-Contra Costa Transit District
5            Alaska Community Foundation
6     Alaska Retirement Management Board
7              Alexander & Baldwin, Inc.
8                         Allergan, Inc.
9       Alyeska Pipeline Service Company

In [4]:
df2.head(10)

Firm Name
0  AAA Northern California, Nevada & Utah Auto Ex...
1                            ACCO Engineered Systems
2                       Adams County Retirement Plan
3                               Adidas America, Inc.
4                                Adobe Systems, Inc.
5                       Advanced Micro Devices, Inc.
6                       AECOM Technology Corporation
7                                    Aera Energy LLC
8                  Aerojet Rocketdyne Holdings, Inc.
9                         Agilent Technologies, Inc.

In [5]:
print(df1.shape)
print(df2.shape)

(266, 1)
(368, 1)


You can see that the company ***AECOM*** appears as ``AECOM`` in CSV1, but as ``AECOM Technology Corporation`` in CSV2.

To do so, we're going to use an external library called ``thefuzz`` that contains functionality to compute the Levenshtein distance between two strings. More on this later.

Your job will be cleaning this data and matching the columns in both CSVs. Let's get started.

## Data Preprocessing

Before we can compare the company names, we need to read the data and align the companies for comparison. To do so, we'll need to use the ``itertools.product`` function to create the "cartesian product" between the two CSVs. Very briefly, the Cartesian Product of two coll
ections ``A`` and ``B``, returns a combination of ALL the elements in both ``A`` and ``B``.

Example:
```py
>>> A = ["Apple", "Alphabet", "Microsoft"]
>>> B = ["MSFT", "Alphabet/Google", "Apple inc."]

>>> list(itertools.product(A, B))
[('Apple', 'MSFT'),
 ('Apple', 'Alphabet/Google'),
 ('Apple', 'Apple inc.'),
 ('Alphabet', 'MSFT'),
 ('Alphabet', 'Alphabet/Google'),
 ('Alphabet', 'Apple inc.'),
 ('Microsoft', 'MSFT'),
 ('Microsoft', 'Alphabet/Google'),
 ('Microsoft', 'Apple inc.')]

```
As you can see, every element in A is matched with every element in B.

In [6]:
## Para el lanzamiento de dos dados la combinatoria sería de la siguiente manera

dado_1 =  [1,2,3,4,5,6]
dado_2 =  [1,2,3,4,5,6]

len(list(itertools.product(dado_1, dado_2)))


36

> **NOTA** ``itertools.product``: Funciona para calcular las posibles combinaciones usando un producto cartesiano de dos series

1. Create the ``df`` dataframe containing the product of the two CSVs

    We have already read the 2 CSVs into the ``df1`` and ``df2`` variables. Now, use the ``itertools.product`` method to create a resulting dataframe ``df`` that will contain the product of the two CSVs. The columns should be named ``CSV 1`` and ``CSV 2``.

    As we have ``266`` rows in ``df1`` and ``368`` in ``df2``, the resulting ``df`` will have ``97,888`` rows (266 * 368), and it'll look something like:

In [7]:
lista = list(itertools.product(list(df1.iloc[:,0]), list(df2.iloc[:,0])))

df = pd.DataFrame(lista, columns=['CSV 1', 'CSV 2'])

Otra solución

In [ ]:
csv_1 = df1['CLIENT'].values
csv_2 = df2['Firm Name'].values

companies = itertools.product(csv_1, csv_2)

df = pd.DataFrame(companies, columns=['CSV 1', 'CSV 2'])

## Calculating the Levenshtein distance

We're going to use the ``partial_ratio`` function from the ``fuzz`` module to compute the "distance" between two strings. The result is a number between 0 and ``100``, with ``100`` indicating a "perfect" match. Please keep in mind that this method is not perfect, as shown in the examples about ``Microsoft`` in the notebook.

Now, we will learn how to calculate the Levenshtein distance between two strings. Here we will user ``partial_ratio`` function from the fuzz module to compute the "ratio" between two strings. The result is a number between 0 and 100, with 100 indicating a "perfect" match. Please note that ``partial_ratio`` gives ratio of the shortest string length to the longest string length. For example, if the first string is ``ABC`` and the second string is ``ABDC``, then the ratio will be 4/5 = 0.80.

In [8]:
fuzz.partial_ratio("Apple", "Apple Inc.")

100

In [9]:
fuzz.partial_ratio("Microsoft", "Apple Inc.")

11

In [10]:
fuzz.partial_ratio("Microsoft", "MSFT")

25

If we have list of strings, we can calculate the Levenshtein distance between each pair of strings in the list.

In [11]:
A = ["Apple", "Alphabet", "Microsoft"]
B = ["MSFT", "Alphabet/Google", "Apple inc."]

Below, we combined the two list ``A`` and ``B`` into a list of tuples ``companies`` using product function from ``itertools`` module.

Then, we calculated the partial ratio for each pair of strings in the list companies using ``partial_ratio`` function from ``fuzz``.

In [12]:
companies = list(itertools.product(A, B))
companies

[('Apple', 'MSFT'),
 ('Apple', 'Alphabet/Google'),
 ('Apple', 'Apple inc.'),
 ('Alphabet', 'MSFT'),
 ('Alphabet', 'Alphabet/Google'),
 ('Alphabet', 'Apple inc.'),
 ('Microsoft', 'MSFT'),
 ('Microsoft', 'Alphabet/Google'),
 ('Microsoft', 'Apple inc.')]

In [13]:
for c1, c2 in companies:
    ratio = fuzz.partial_ratio(c1, c2)
    print(f"{c1} > {c2}: {ratio}")

Apple > MSFT: 0
Apple > Alphabet/Google: 40
Apple > Apple inc.: 100
Alphabet > MSFT: 0
Alphabet > Alphabet/Google: 100
Alphabet > Apple inc.: 38
Microsoft > MSFT: 25
Microsoft > Alphabet/Google: 22
Microsoft > Apple inc.: 22


You will see the greater the ratio, the more similar the strings are.

2. Create a new column ``Ratio Score`` that contains the distance for all the rows in df

    Now apply the function ``fuzz.partial_ratio`` to all the companies in ``df`` to calculate the distance between them. Store the distance in a new column named ``Ratio Score``. It'll look similar to:

For the simplest solution, we can use list comprehension to generate the scores from the dataframe ``df``:

In [ ]:
scores = [fuzz.partial_ratio(c1, c2) for c1, c2 in df.values]
df['Ratio Score'] = scores

Or, we can use the apply method to apply the function to all the rows in df:

In [15]:
df['Ratio Score'] = df.apply(lambda row: fuzz.partial_ratio(row['CSV 1'], row['CSV 2']), axis=1)

In [16]:
df.values[0]

array(['Adobe Systems, Inc.',
       'AAA Northern California, Nevada & Utah Auto Exchange', 26],
      dtype=object)

3. How many rows have a Ratio score of 90 or more?

In [21]:
len(df[df['Ratio Score'] >= 90])

106

4. What's the corresponding company in CSV2 to ``AECOM`` in CSV1?

    We saw that in ``CSV 1`` there's a company ``AECOM``, what's the corresponding value in ``CSV 2``? The higher the Ratio Score, the more likely it is to be a match. So, use 

In [25]:
df[(df['CSV 1'] == 'AECOM') & (df['Ratio Score'] > 90)]['CSV 2']

742    AECOM Technology Corporation
Name: CSV 2, dtype: object

5. What's the corresponding ``CSV 2`` company of Starbucks?

    ``CSV 1`` company is ``Starbucks``, what's the corresponding name in ``CSV 2``?

In [26]:
df[(df['CSV 1'] == 'Starbucks') & (df['Ratio Score'] > 90)]['CSV 2']

77948    Starbucks Corporation
Name: CSV 2, dtype: object

6. Is there a matching company for ``Pinnacle West Capital Corporation`` in column ``CSV 2``?

    Column ``CSV 1`` contains ``Pinnacle West Capital Corporation``, name a matching company in column ``CSV 2`` with a ratio score more than ``90``?

In [29]:
df[(df['CSV 1'] == 'Pinnacle West Capital Corporation') & (df['Ratio Score'] >= 80)]

CSV 1               CSV 2  Ratio Score
61128  Pinnacle West Capital Corporation  Avista Corporation           83
61130  Pinnacle West Capital Corporation    Ball Corporation           88

7. How many matching companies are there for ``County of Los Angeles Deferred Compensation Program``?

    ``CSV 1`` contains the ``County of Los Angeles Deferred Compensation Program``. How many matching companies seem to be in ``CSV 2``?

In [31]:
df[(df['CSV 1'] == 'County of Los Angeles Deferred Compensation Program') & (df['Ratio Score'] > 90)]

CSV 1  \
26206  County of Los Angeles Deferred Compensation Pr...   
26227  County of Los Angeles Deferred Compensation Pr...   

                                                   CSV 2  Ratio Score  
26206          City of Los Angeles Deferred Compensation           95  
26227  County of Los Angeles Deferred Compensation Pr...          100

8. Is there a matching company for ``The Queens Health Systems``?

    ``CSV 1`` contains ``The Queens Health Systems``, is there a matching in CSV 2?

In [32]:
df[(df['CSV 1'] == 'The Queens Health Systems') & (df['Ratio Score'] > 90)]

CSV 1                       CSV 2  Ratio Score
84220  The Queens Health Systems  The Queen's Health Systems           96

# String pre-processing for Sentiment Analysis in NLP

In this project, we'll use NLTK, the de-facto and most popular NLP Python library to perform some Sentiment Analysis on a corpus of +22K tweets scraped during the 2022 FIFA world cup.

But the objective of this project is not to perform Sentiment Analysis or NLP, as that's outside of the scope of this skill track. The goal of this project is to perform several pre-processing and cleaning tasks to the text of the tweets in order to render them ready for the Sentiment Analysis task.

The dataset we're using already includes the classification of Sentiment which was performed with an advanced Deep Learning model based on on [Meta's AI RoBERTa](https://ai.meta.com/blog/roberta-an-optimized-method-for-pretraining-self-supervised-nlp-systems/).

Our goal will be to perform the right pre-processing tasks and then use NLTK's built-in [VADER](https://www.nltk.org/howto/sentiment.html) module to perform a very rudimentary version of Sentiment Analysis, and compare its performance with the original's Sentiment column included in the dataset.

Most NLP projects require several cleaning and pre-processing tasks, like tokenization, stemming, lemmatization, etc. Don't worry if you don't understand what they are, we'll explain them as they're required in the project. The important part is to understand that the complexity of NLP usually relies mostly on the pre-processing and cleaning, than the NLP models themselves.

Let's get started!

1. Read dataset, but only the ``Tweet`` and ``Sentiment`` columns

    Read the data in ``fifa_world_cup_2022_tweets.csv`` into a dataframe, but only the columns ``Tweet`` and ``Sentiment``.

    Your ``df`` should look something like:

In [4]:
import re

import pandas as pd
import numpy as np

In [34]:
import nltk
nltk.download('punkt')
from nltk.sentiment.vader import SentimentIntensityAnalyzer
from nltk.corpus import stopwords
from nltk.tokenize import word_tokenize
from nltk.stem import WordNetLemmatizer

[nltk_data] Downloading package punkt to
[nltk_data]     C:\Users\DVadeus\AppData\Roaming\nltk_data...
[nltk_data]   Unzipping tokenizers\punkt.zip.


1. Read dataset, but only the Tweet and Sentiment columns

    Read the data in ``fifa_world_cup_2022_tweets.csv`` into a dataframe, but only the columns ``Tweet`` and ``Sentiment``.

    Your ``df`` should look something like:

In [56]:
df = pd.read_csv('files/fifa_world_cup_2022_tweets.csv', usecols =['Tweet', 'Sentiment'])

2. Lowercase the column ``Tweet`` in the new column ``Tweet Lower``

    Create a new column Tweet Lower that contains the contents of the ``Tweet`` column, but all lowercased.

In [57]:
df['Tweet Lower'] = df['Tweet'].str.lower()

In [58]:
print(df.iloc[0]['Tweet Lower'])

what are we drinking today @tucantribe 
@madbears_ 
@lkinc_algo 
@al_goanna 

#worldcup2022 https://t.co/oga3tzvg5h


3. Remove all URLs from ``Tweet Lower`` in ``Tweet Clean``

    URLs, hashtags, mentions are mostly useless elements when it comes to sentiment analysis. We'll start by removing all the URLs. Remove all the URLs from Tweet Lower and store your results in Tweet Clean.

    Warning! Don't forget to remove any leading or trailing whitespaces. For example, if you remove the URL from the following tweet:

    ```py
    what are we drinking today @tucantribe 
    @madbears_ 
    @lkinc_algo 
    @al_goanna 
    
    #worldcup2022 https://t.co/oga3tzvg5h
    ```

    The result should be:
    
    ``` py
    """what are we drinking today @tucantribe 
    @madbears_ 
    @lkinc_algo 
    @al_goanna 

    #worldcup2022"""
    ```
    
    Without a trailing space after the #worldcup2022 hashtag.

    > *Use this pattern for matching the url: ``https?:\/\/[^\s]*``*

In [59]:
df['Tweet Clean'] = df.apply(lambda row: re.sub('https?:\/\/[^\s]*', "", row['Tweet Lower'], flags=re.MULTILINE).strip(), axis=1)

In [60]:
df['Tweet Clean'][0]

'what are we drinking today @tucantribe \n@madbears_ \n@lkinc_algo \n@al_goanna \n\n#worldcup2022'

4. Remove username mentions in ``Tweet Clean``

    Still in ``Tweet Clean``, remove any twitter mentions (in the form ``@datawars_io``). In this case, we're modifying the original column Tweet Clean, so if you make a mistake, you'll have to re-run your previous code and start over.

    Remember to strip any trailing or leading whitespaces.

In [61]:
df['Tweet Clean'] = df['Tweet Clean'].str.replace('@[\w]*', '',regex = True, flags = re.MULTILINE).str.strip()

5. Remove Hashtags from Tweet Clean

    Still in Tweet Clean, remove any hashtags.

In [62]:
df['Tweet Clean'] = df['Tweet Clean'].str.replace('#[\w]*', '',regex = True, flags = re.MULTILINE).str.strip()

## Tokenizing Tweets

6. Create the list ``tokenized_tweets`` by applying the function ``word_tokenize`` to the values of the column ``Tweet Clean``

    We'll now start using the ``nltk`` module. Don't worry if you've never used it before, as these are all simple functions that don't require an NLP background.

    We'll start by "tokenizing" the tweets. Tokenizing means basically splitting a corpus of text into different words or tokens.

    Your task is to use the ``word_tokenize`` function to create a list of tweet tokens and store the result in ``tokenized_tweets``. This means that ``tokenized_tweets`` is a list of lists, a list of tokens in the following form:
    ```py
    [
        ['what', 'are', 'we', 'drinking', 'today'], # tweet
        ['worth', 'reading', 'while', 'watching'],  # tweet
    ]
    ```

In [64]:
tokenized_tweets = [word_tokenize(w) for w in df['Tweet Clean'].values]

7. Filter stop words

    Stop words are words that don't contribute much to the meaning of a sentence, like conjunctions ("for", "and") or the word "the", "a", etc. The ``nltk`` module contains stop words for english, that we can get with ``stopwords.words('english')``.

    Your task is to remove any stop words from the tokens you have previously generated. Store your results in the variable ``filtered_tokenized_tweets``, which continues to be a list of lists, but with the stop words filtered out.

In [42]:
nltk.download('stopwords')

[nltk_data] Downloading package stopwords to
[nltk_data]     C:\Users\DVadeus\AppData\Roaming\nltk_data...
[nltk_data]   Unzipping corpora\stopwords.zip.


True

In [53]:
stop_words_english = set(stopwords.words('english'))

In [69]:
filtered_tokenized_tweets = [[w for w in tweet if w not in stop_words_english] for tweet in tokenized_tweets]

## Performing Sentiment Analysis

We'll now start the process of sentiment analysis using the builtin ``SentimentIntensityAnalyzer`` NLTK class, also known as VADER.

Vader works well for the basic tasks, for example this will be classified mostly as a positive sentence:

```py
>>> analyzer.polarity_scores("DataWars is awesome! I love it so much!")
{'neg': 0.0, 'neu': 0.36, 'pos': 0.64, 'compound': 0.8715}
```

But it can't handle things like sarcasm or rethorical questions. For example, sarcasm is completely lost to the analyzer, this is clearly a negative sentence:

```py
>>> analyzer.polarity_scores("Yeah, R is sure a great programming language. If you are 80 years old")
{'neg': 0.0, 'neu': 0.511, 'pos': 0.489, 'compound': 0.8225}
```

While the following one, which is not as extreme as the previous one, could also be arguably classified as negative. Instead, it's found to be completely neutral:
```py
>>> analyzer.polarity_scores("I can't believe how people still program in R in 2019, what is going on?")
{'neg': 0.0, 'neu': 1.0, 'pos': 0.0, 'compound': 0.0}
```

Don't worry too much about the contents of the resulting dictionary, we'll explain them in the following activities.

8. Glue all the tweets back again

    Use a single space to concat the tokens that we have preprocessed in our previous tasks and build the tweet again. Store your results in the variable ``cleaned_tweets``. In this case, it'll no longer be a list of lists, but a list of strings, the tweets we have assembled again, and it'll look something like:
    ```py
    ['drinking today',
    'amazing launch video . shows much face canada men ’ national team changed since last world cup entry 1986. ’ wait see boys action ! canada : fifa world cup opening video',
    'worth reading watching']
    ```

In [83]:
cleaned_tweets = []
for tweet in filtered_tokenized_tweets:
    word = ''
    for t in tweet:
        word += t+' '
    cleaned_tweets.append(word.strip())

Mejor solución

In [85]:
cleaned_tweets = [' '.join(tweet) for tweet in filtered_tokenized_tweets]
cleaned_tweets[:5]

['drinking today',
 'amazing launch video . shows much face canada men ’ national team changed since last world cup entry 1986. ’ wait see boys action ! canada : fifa world cup opening video',
 'worth reading watching',
 'golden maknae shinning bright',
 'bbc cares much human rights , homosexual rights , women rights say opening ceremony ? ? saying opening ceremony ? ? bbc censor opening ceremony ? ?']

9. Apply VADER to all the words

    Use the ``analyzer.polarity_scores`` method to perform sentiment analysis on all the tweets in ``cleaned_tweets``. Store the list of results in the variable ``tweet_sentiment_scores``.

    As we mentioned before, this requires just a method invocation:
    ```py
    >>> analyzer.polarity_scores(YOUR_TWEET)
    ```
    Your ``tweet_sentiment_scores`` variable will look something like:
    ```py
    [
        {'neg': 0.0, 'neu': 1.0, 'pos': 0.0, 'compound': 0.0},
        {'neg': 0.0, 'neu': 0.864, 'pos': 0.136, 'compound': 0.6239},
        {'neg': 0.0, 'neu': 0.513, 'pos': 0.487, 'compound': 0.2263},
        ...
    ]
    ```

In [92]:
nltk.download('vader_lexicon')

[nltk_data] Downloading package vader_lexicon to
[nltk_data]     C:\Users\DVadeus\AppData\Roaming\nltk_data...


True

In [93]:
analyzer = SentimentIntensityAnalyzer()

In [96]:
tweet_sentiment_scores = [analyzer.polarity_scores(tweet) for tweet in cleaned_tweets]

In [97]:
tweet_sentiment_scores[:5]

[{'neg': 0.0, 'neu': 1.0, 'pos': 0.0, 'compound': 0.0},
 {'neg': 0.0, 'neu': 0.864, 'pos': 0.136, 'compound': 0.6239},
 {'neg': 0.0, 'neu': 0.513, 'pos': 0.487, 'compound': 0.2263},
 {'neg': 0.0, 'neu': 0.508, 'pos': 0.492, 'compound': 0.4404},
 {'neg': 0.13, 'neu': 0.739, 'pos': 0.13, 'compound': 0.0}]

10. Calculate the sentiment of each tweet based on the following rule....

    The result of ``analyzer.polarity_scores`` is a dictionary with several keys:
    
    ```py
    >>> analyzer.polarity_scores("DataWars is awesome! I love it so much!")
    {'neg': 0.0, 'neu': 0.36, 'pos': 0.64, 'compound': 0.8715}
    ```
    
    The ``neg``, ``neu`` and ``pos`` keys represent the proportions of the text that fall in each category (Negative, Neutral and Positive). They add up to 1.

    But the key that we're really interested in is ``compound``, which is a weighted composite score that has been normalized between -1 (most extreme negative) and +1 (most extreme positive). In this case, the ``compound`` score of 0.8715 indicates a very high positive sentiment.

    The general rule of thumb for interpreting the ``compound`` score is:

    - Positive sentiment: ``compound`` score > 0.05
    - Neutral sentiment: ``compound`` score between -0.05 and 0.05
    - Negative sentiment: ``compound`` score < -0.05
    
    Calculate the sentiment of each score and store it in the variable ``tweet_sentiment_results`` that should look something like: ``['neutral', 'positive', 'positive', ...]``.

In [100]:
tweet_sentiment_results = ['positive' if score['compound'] > 0.05 else 'negative' if score['compound'] < -0.05 else 'neutral' for score in tweet_sentiment_scores]

Otra opción

In [ ]:
def get_sentiment(res):
    comp = res['compound']
    if comp > .05:
        return "positive"
    elif comp < -.05:
        return "negative"
    return "neutral"


In [ ]:
tweet_sentiment_results = [get_sentiment(score) for score in tweet_sentiment_scores]
tweet_sentiment_results[:3]

11. Delete the columns ``Tweet Lower`` and ``Tweet Clean`` from and add the new column ``Calculated Sentiment``

    Remove the columns we previously used (``Tweet Lower``, ``Tweet Clean``) and create a new one named ``Calculated Sentiment`` with the results of ``tweet_sentiment_results``.

In [102]:
df.drop(columns=['Tweet Lower', 'Tweet Clean'], inplace = True)
df['Calculated Sentiment'] = tweet_sentiment_results

12. How many tweets were incorrectly classified?

    Assuming the column ``Sentiment`` had the correct sentiment, how many did we classified erroneously in our ``Calculated Sentiment`` column?

In [104]:
df[df['Sentiment'] != df['Calculated Sentiment']]

Tweet Sentiment  \
4      If the BBC cares so much about human rights, h...  negative   
6      Look like a only me and the Jamaican football ...   neutral   
7      Really? Football on a Monday morning at 9 and ...  negative   
8      As the World Cup starts in Qatar, it’s Black A...  positive   
10     #Qatar tried to help its useless soccer team 5...  negative   
...                                                  ...       ...   
22514  #WorldCup2022 is starting tonight\nAnd with th...  negative   
22517  The World Cup starts TODAY!\n\n#Qatar2022 #Foo...  positive   
22519            Here We go World cup 2022 #WorldCup2022  positive   
22520  Anderlecht confirms former Viborg FF's Jesper ...   neutral   
22523  How to buy $SOT on PinkSale?🤔\n\nHave you been...   neutral   

      Calculated Sentiment  
4                  neutral  
6                 positive  
7                  neutral  
8                 negative  
10                positive  
...                    ...  
22514             positive  
22517              neutral  
22519              neutral  
22520             positive  
22523             negative  

[9508 rows x 3 columns]

13. How many Negative tweets were incorrectly classified (either as ``Neutral`` or ``Positive``)?

In [105]:
df[(df['Sentiment'] == 'negative') & (df['Sentiment'] != df['Calculated Sentiment'])]

Tweet Sentiment  \
4      If the BBC cares so much about human rights, h...  negative   
7      Really? Football on a Monday morning at 9 and ...  negative   
10     #Qatar tried to help its useless soccer team 5...  negative   
36     If the BBC cares so much about human rights, g...  negative   
38     With your looks, you couldn't even get with a ...  negative   
...                                                  ...       ...   
22448  If the Royal Family can pressure the Supreme C...  negative   
22449  I can’t believe the World Cup starts tomorrow....  negative   
22455  The World Cup should have started today so the...  negative   
22496  #Qatar doing to the #WorldCup2022 what @elonmu...  negative   
22514  #WorldCup2022 is starting tonight\nAnd with th...  negative   

      Calculated Sentiment  
4                  neutral  
7                  neutral  
10                positive  
36                 neutral  
38                 neutral  
...                    ...  
22448             positive  
22449              neutral  
22455              neutral  
22496             positive  
22514             positive  

[3026 rows x 3 columns]

# Practicing identifyin and dealing with invalid values

In this project, we will be working with a Human Resources dataset containing information about employees from a fictional company. The dataset includes details such as employee ID, date of birth, gender, marital status, job title, department, salary, performance score, and other columns.

Invalid or missing values are common issues that can arise in datasets like these, and it's important to be able to identify and clean these invalid values to avoid biased or inaccurate analyses. In this lab, we will be focusing on cleaning invalid values in selected columns using different methods in pandas.

We will start by exploring the dataset and identifying any invalid or missing values in the selected columns using different methods to clean the data, based on the data type. We will focus on 4 different types: - Invalid numeric column - Invalid categories - Invalid datetimes - Invalid patterns in strings

By the end of this lab, you should feel comfortable using pandas to clean invalid values in a dataset, and have a better understanding of how to use different methods and techniques to achieve the desired outcome. The lab includes basic activities and advanced activities, focusing on different aspects of cleaning invalid values in the Human Resources dataset.

In [1]:
import pandas as pd

In [2]:
df = pd.read_csv('files/HRDataset.csv')

In [3]:
df.head()

Employee_Name  EmpID  DeptID  Salary  PositionID  \
0       Adinolfi, Wilson  K  10026       5   62506          19   
1  Ait Sidi, Karthikeyan     10084       3  104437          27   
2         Akinkuolie, Sarah  10196       5   64955          20   
3              Alagbe,Trina  10088       5   64991          19   
4          Anderson, Carol   10069       5   50825          19   

                   Position State   Zip       DOB Sex  ... RecruitmentSource  \
0   Production Technician I    MA  1960  07/10/83   M  ...          LinkedIn   
1                   Sr. DBA    MA  2148  05/05/75   M  ...            Indeed   
2  Production Technician II    MA  1810  09/19/88   F  ...          LinkedIn   
3   Production Technician I    MA  1886  09/27/88   F  ...            Indeed   
4   Production Technician I    MA  2169  09/08/89   F  ...     Google Search   

  PerformanceScore EngagementSurvey EmpSatisfaction SpecialProjectsCount  \
0          Exceeds             4.60               5                    0   
1      Fully Meets             4.96               3                    6   
2      Fully Meets             3.02               3                    0   
3      Fully Meets             4.84               5                    0   
4      Fully Meets             5.00               4                    0   

  LastPerformanceReview_Date DaysLateLast30 Absences                   Email  \
0                  1/17/2019              0        1    Adinolfi@company.com   
1                  2/24/2016              0       17    Ait Sidi@company.com   
2                  5/15/2012              0        3  Akinkuolie@company.com   
3                   1/3/2019              0       15      Alagbe@company.com   
4                   2/1/2016              0        2    Anderson@company.com   

               Phone  
0   +62 199-807-6651  
1  +358 128-157-9500  
2  +474 737-382-6683  
3  +806 538-458-2517  
4  +148 658-824-9500  

[5 rows x 31 columns]

In [4]:
df.head().T

0                         1  \
Employee_Name                   Adinolfi, Wilson  K  Ait Sidi, Karthikeyan      
EmpID                                         10026                     10084   
DeptID                                            5                         3   
Salary                                        62506                    104437   
PositionID                                       19                        27   
Position                    Production Technician I                   Sr. DBA   
State                                            MA                        MA   
Zip                                            1960                      2148   
DOB                                        07/10/83                  05/05/75   
Sex                                               M                         M   
MaritalDesc                                  Single                   Married   
CitizenDesc                              US Citizen                US Citizen   
HispanicLatino                                   No                        No   
RaceDesc                                      White                     White   
DateofHire                                 7/5/2011                 3/30/2015   
DateofTermination                               NaN                 6/16/2016   
TermReason                        N/A-StillEmployed             career change   
EmploymentStatus                             Active    Voluntarily Terminated   
Department                        Production                            IT/IS   
ManagerName                          Michael Albert                Simon Roup   
ManagerID                                      22.0                       4.0   
RecruitmentSource                          LinkedIn                    Indeed   
PerformanceScore                            Exceeds               Fully Meets   
EngagementSurvey                                4.6                      4.96   
EmpSatisfaction                                   5                         3   
SpecialProjectsCount                              0                         6   
LastPerformanceReview_Date                1/17/2019                 2/24/2016   
DaysLateLast30                                    0                         0   
Absences                                          1                        17   
Email                          Adinolfi@company.com      Ait Sidi@company.com   
Phone                              +62 199-807-6651         +358 128-157-9500   

                                                   2                        3  \
Employee_Name                      Akinkuolie, Sarah             Alagbe,Trina   
EmpID                                          10196                    10088   
DeptID                                             5                        5   
Salary                                         64955                    64991   
PositionID                                        20                       19   
Position                    Production Technician II  Production Technician I   
State                                             MA                       MA   
Zip                                             1810                     1886   
DOB                                         09/19/88                 09/27/88   
Sex                                                F                        F   
MaritalDesc                                  Married                  Married   
CitizenDesc                               US Citizen               US Citizen   
HispanicLatino                                    No                       No   
RaceDesc                                       White                    White   
DateofHire                                  7/5/2011                 1/7/2008   
DateofTermination                          9/24/2012                      NaN   
TermReason                                     hours        N/A-StillEmployed   

## Discovering the dataset

In [5]:
#show the number of cells in the dataframe
print("dataset size: ", df.size)

#show the number of records (rows) in the dataframe
print("number of employees: ", len(df))

#show the number of features (columns) in the dataframe
print("number of features: ", len(df.columns)) 

dataset size:  9641
number of employees:  311
number of features:  31


1. Clean the column Salary by removing invalid values

    Invalid values are defined as any value that is not an integer.

    Perform the selection of valid values and store them in column ``Salary_Fixed`` while invalid values should be ``NaN``. Then select invalid values and store the results in the variable ``df_invalid_salaries``.

In [6]:
df['Salary_Fixed'] = df[df['Salary'].str.isnumeric()]['Salary']

In [7]:
df_invalid_salaries = df.loc[df['Salary_Fixed'].isna(), ['Salary']]

In [8]:
df_invalid_salaries

Salary
22   oFtFtkls
41   HEbELLmj
62   DMZIQhGz
71   XSfelyZh
117  RxhGvEBX
138  xqCStzCF
165  gHNxOVWk
216  WXrOyJEO
220  OAWNTcGQ
259  dvCsKwyl
291  ZGEoVKAo
304  KFpwkWHX

2. Clean the column ``Zip`` by removing invalid values

    Invalid values are defined as any value that is not an integer.

    Perform the selection of valid values and store them in column ``Zip_Fixed`` while invalid values should be ``NaN``. Then select invalid values and store the results in the variable ``df_invalid_zip``.

In [9]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 311 entries, 0 to 310
Data columns (total 32 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   Employee_Name               311 non-null    object 
 1   EmpID                       311 non-null    int64  
 2   DeptID                      311 non-null    int64  
 3   Salary                      311 non-null    object 
 4   PositionID                  311 non-null    int64  
 5   Position                    311 non-null    object 
 6   State                       311 non-null    object 
 7   Zip                         311 non-null    object 
 8   DOB                         310 non-null    object 
 9   Sex                         311 non-null    object 
 10  MaritalDesc                 311 non-null    object 
 11  CitizenDesc                 311 non-null    object 
 12  HispanicLatino              311 non-null    object 
 13  RaceDesc                    311 non

In [15]:
df['Zip_Fixed'] = df.loc[df['Zip'].str.isnumeric(), ['Zip']]

df_invalid_zip = df.loc[df['Zip_Fixed'].isna(), ['Zip']]

0      1960
1      2148
2      1810
3      1886
4      2169
       ... 
306    1810
307    2458
308    2067
309    2148
310    1730
Name: Zip_Fixed, Length: 311, dtype: object

3. Clean the column ``ManagerID`` by removing invalid values.

    Invalid values are defined as any value that is not an integer or is an integer below 1.

    Perform the selection of valid values and store them in column ``ManagerID_Fixed`` while invalid values should be ``NaN``. Then select invalid values and store the results in the variable ``df_invalid_managerID``.

In [24]:
df['ManagerID'] = pd.to_numeric(df['ManagerID'], errors = 'coerce')

In [21]:
df['ManagerID_Fixed'] = pd.to_numeric(df['ManagerID'], errors = 'coerce')

In [22]:
df_invalid_managerID = df.loc[df['ManagerID_Fixed'].isna(),['ManagerID']]

In [23]:
df_invalid_managerID

ManagerID
0        22.0
1         4.0
2        20.0
3        16.0
4          No
..        ...
306      20.0
307      12.0
308       2.0
309       4.0
310      14.0

[310 rows x 1 columns]